### PR TITLE
Unit test

### PR DESCRIPTION
**Pull request by destromas1**

### CODE REVIEW

**Review by knutole — APPROVED** [review]
Excellent stuff! 👍

**Comment by destromas1** [comment]
Thanks :) 👍

File: js/src/models/model.project.js [hunk 1]
```diff
@@ -348,10 +348,19 @@ Wu.Model.Project = Wu.Model.extend({
...
+		this.store.name = result.project.name;
+
+		// update slug name
+		!window.testMode && this.setSlug(result.project.name);
+
 		Wu.Mixin.Events.fire('projectChanged', { detail : {
 			projectUuid : this.getUuid(),
 			name : result.project.name
-		}});
+		}, projectChangedFired : true});
```
Comment (knutole) [inline js/src/models/model.project.js]: why is this necessary?

Comment (destromas1) [reply js/src/models/model.project.js]: While I was working , I saw that this event was being fired without  calling ```Wu.Mixin.Events.fire('projectChanged'```
So I tried to specify the path I want to fire. I am looking more whats the other side.

**Comment by destromas1** [comment]
@knutole https://github.com/mapic/mapic.js/issues/21 Bug is also fixed by the commit https://github.com/mapic/mapic.js/pull/18/commits/9bb513ae55cafa31d1a121793f21611615828aec

File: js/src/models/model.project.js (hunk 1 above)
Comment (destromas1) [reply js/src/models/model.project.js]: https://github.com/mapic/mapic.js/blob/master/js/src/panes/pane.map.js#L260-L268

Here this event was called  2 times ,  But as the event was not firing before , so this 2 calls were just there , but not executing.

File: js/src/chrome/projects/chrome.projects.js [hunk 2]
```diff
@@ -1255,7 +1257,7 @@ Wu.Chrome.Projects = Wu.Chrome.extend({
 		}
 
 		// slack
-		this._logUpdate({
+		!window.testMode && this._logUpdate({
```
Comment (knutole) [inline js/src/chrome/projects/chrome.projects.js]: This can be moved to the `app.log()` function, so you dont have to put `if (window.testMode)` in so many places. Better to put them in the most common function, ie. `app.log()`

File: js/src/chrome/projects/chrome.projects.js [hunk 3]
```diff
@@ -1346,6 +1348,8 @@ Wu.Chrome.Projects = Wu.Chrome.extend({
 			return;
 		}
 
+		if(window.testMode) return true;
```
Comment (knutole) [inline js/src/chrome/projects/chrome.projects.js]: Why is this here exactly?

File: js/src/chrome/projects/chrome.projects.js [hunk 4]
```diff
@@ -1458,6 +1463,22 @@ Wu.Chrome.Projects = Wu.Chrome.extend({
 		app.Controller.openFirstProject();
 		
 	},
+
+	_onProjectChanged : function (e) {
```
Comment (knutole) [inline js/src/chrome/projects/chrome.projects.js:1507]: Why was is necessary to add this?

File: js/src/chrome/users/chrome.users.js [hunk 5]
```diff
@@ -454,6 +454,8 @@ Wu.Chrome.Users = Wu.Chrome.extend({
 	// slack, analytics
 	_logInvites : function (options) {
 
+		if(window.testMode)	return;
```
Comment (knutole) [inline js/src/chrome/users/chrome.users.js]: Same as above, should be moved to `app.log`

File: js/src/models/model.project.js [hunk 6]
```diff
@@ -348,6 +348,15 @@ Wu.Model.Project = Wu.Model.extend({
...
 			description : result.error
 		});
+
+		if(result.updated[0] == "slug") return;
+
+		// store on server
+		this.store.name = result.project.name;
+
+		// update slug name
+		!window.testMode && this.setSlug(result.project.name);
```
Comment (knutole) [inline js/src/models/model.project.js]: `slug` and `name` are two different things.. What's going on here? Is `slug` returned as a `name` key?

File: js/src/panes/pane.map.js [hunk 7]
```diff
@@ -256,18 +256,19 @@ Wu.MapPane = Wu.Pane.extend({
     },
 
     _onMove : function () {
```
Comment (knutole) [inline js/src/panes/pane.map.js:258]: Does this `_onMove` function ever fire?

Why is it necessary to comment out the event fire?

File: test/core/project.spec.js [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+  it("should update project name", function () {
+        
+    var store = {"name":"Test Project Name","uuid" : "project-0e386d2a-2966-419b-8604-96d112d4abb2","description":"Test Project description","createdByName":"Shahjada Talukdar","access":{"edit":[],"read":[],"options":{"share":true,"download":false,"isPublic":false}}};
+    
+    var project = new Wu.Model.Project(store);
+
+    var newProjectName = "Test Project Name 2";
+
+    project.setName(newProjectName);
+
```
Comment (knutole) [inline test/core/project.spec.js:97]: Missing `expect`?

File: test/core/project.spec.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+  });
+
+  it("should delete a project", function () {
+        
+    var store = {"name":"Test Project Name","uuid" : "project-0e386d2a-2966-419b-8604-96d112d4abb2","description":"Test Project description","createdByName":"Shahjada Talukdar","access":{"edit":[],"read":[],"options":{"share":true,"download":false,"isPublic":false}}};
+
+    var project = new Wu.Model.Project(store);
+
+    project._delete(function (err ,resp) {
+      expect(resp.deleted).to.be.true;
```
Comment (knutole) [inline test/core/project.spec.js]: Could also check that project is no longer in `app.Projects`

File: test/core/project.spec.js [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+      value : ""
+    };
+
+    users._access = {
+      edit : [],
+      read :[]
+    };
+
+    users._sendInvites({target : {}});
+
```
Comment (knutole) [inline test/core/project.spec.js:146]: missing `expect`?

**Comment by knutole** [comment]
@destromas1 Could you please reply on all comments in review? :)

File: js/src/panes/pane.map.js (hunk 7 above)
Comment (destromas1) [reply js/src/panes/pane.map.js:258]: Yes `onMove` and `onZoom` fire , when we move / zoom the Map in UI.

File: test/core/project.spec.js (hunk 8 above)
Comment (destromas1) [reply test/core/project.spec.js:97]: Yes `expect` missing , as this has event based coding , not function callback. We need to get a way for this. Will talk to you.

File: test/core/project.spec.js (hunk 9 above)
Comment (destromas1) [reply test/core/project.spec.js]: Sure , will check.

File: test/core/project.spec.js (hunk 10 above)
Comment (destromas1) [reply test/core/project.spec.js:146]: Yes `expect` missing , as this has event based coding , not function callback. We need to get a way for this. Will talk to you.

**Comment by destromas1** [comment]
@knutole Replied 👍

File: test/core/project.spec.js (hunk 9 above)
Comment (knutole) [reply test/core/project.spec.js]: I mean adding check to test :)

File: js/src/panes/pane.map.js (hunk 7 above)
Comment (knutole) [reply js/src/panes/pane.map.js:258]: Why is it necessary to comment out the event fire?

**Comment by knutole** [comment]
Missing quite a few comments @destromas1

File: js/src/chrome/projects/chrome.projects.js (hunk 2 above)
Comment (destromas1) [reply js/src/chrome/projects/chrome.projects.js]: Yup , already in mind and queue . Thanks for the notice too!

File: js/src/panes/pane.map.js (hunk 7 above)
Comment (destromas1) [reply js/src/panes/pane.map.js:258]: Coz when we move and /or Zoom , Project Update Functions should not be called , right ? Project Update function updates Project Name and Slug . Then why in every Mouse Move or Zoom , we need to do this heavy Server Call as we are not actually wanting to update Name and Slug

And the Event was not getting fired before at all , only the Code was there , but as the event had issue , it was doing nothing!

File: test/core/project.spec.js (hunk 9 above)
Comment (destromas1) [reply test/core/project.spec.js]: Yup , Yup , i also meant so :)

File: js/src/chrome/projects/chrome.projects.js (hunk 3 above)
Comment (knutole) [reply js/src/chrome/projects/chrome.projects.js]: @destromas1

File: js/src/chrome/projects/chrome.projects.js (hunk 4 above)
Comment (knutole) [reply js/src/chrome/projects/chrome.projects.js:1507]: @destromas1

File: js/src/models/model.project.js (hunk 6 above)
Comment (knutole) [reply js/src/models/model.project.js]: @destromas1 Slug and Name are two different things. Project name can be "Test Project", but `slug` is the project name that is in the browser url, so it would be `testproject` for example.  So you should put `setSlug(project.name`), because if name is "Test Project" the url will be broken.. That's why we use slugs. See the difference?

File: js/src/panes/pane.map.js (hunk 7 above)
Comment (knutole) [reply js/src/panes/pane.map.js:258]: Firing `projectChanged` doesn't necessarily update project name or slug, does it? I mean, it shouldn't. It's just an event telling the system that project has changed (could be position, could be name, could be settings). So I'm very surprised if firing `projectChanged` sends a server request with project name. 

If so, this has to be fixed, because that's just wrong logic.

Comment (knutole) [reply js/src/panes/pane.map.js:258]: @destromas1

Comment (destromas1) [reply js/src/panes/pane.map.js:258]: https://github.com/mapic/mapic.js/blob/master/js/src/models/model.project.js#L357
This previous event was actually calling API for `setSlug` . But as we know now , this event was not get called . So I think we need to think and refactor this Update Project Flow. Like - separating - Update Project Name +details and Slug. I will think and come up with a refactor

Comment (knutole) [reply js/src/panes/pane.map.js:258]: Right, thanks for that. This is definitively wrong. I see now it was someone else who wrote it, that's why I'm confused. The whole function should be commented out. That's the immediate fix. Then probably there will be an issue with name not saving, but we'll solve that after. Can you comment it out now, first?

Comment (knutole) [reply js/src/panes/pane.map.js:258]: There are basically two errors in that code: 1) We shouldn't save name on every `projectChanged`, because it could be something else that changed. 2) `.setSlug()` shouldnt be called with `project.name`, because name & slug are two different things.

**Comment by knutole** [comment]
@destromas1 This is very nice work, Shah:) Just a lot of details we have to make sure are correct!

File: js/src/chrome/projects/chrome.projects.js (hunk 4 above)
Comment (destromas1) [reply js/src/chrome/projects/chrome.projects.js:1507]: @knutole  current architecture is like > We fire Events from Model Layer and receive the fired Event inside Chrome layer . Exm-> https://github.com/mapic/mapic.js/blob/master/js/src/models/model.project.js#L440 this fires and which is received at https://github.com/mapic/mapic.js/blob/master/js/src/chrome/projects/chrome.projects.js#L1451

And the previous ProjectChanged Event was never got fired , it was in model layer.

File: js/src/models/model.project.js (hunk 6 above)
Comment (destromas1) [reply js/src/models/model.project.js]: @knutole Yup I got that and refactored code and separated Project Update and Slug Update

File: test/core/project.spec.js (hunk 8 above)
Comment (destromas1) [reply test/core/project.spec.js:97]: Done.

**Comment by knutole** [comment]
@destromas1 What's the status on this PR?

**Comment by destromas1** [comment]
@knutole 
U can review the PR and merge if you feel okay and plz give ur feedback!

File: js/src/core/app.js [hunk 11]
```diff
@@ -32,7 +32,8 @@ Wu.App = Wu.Class.extend({
 	},
 
 	_raven : function () {
-		if(window.testMode) return true;
+		if(window.testMode || location.hostname === "localhost") return true;
```
Comment (knutole) [inline js/src/core/app.js]: Isn't there another way to do this? I mean, to get the logs AND Raven on localhost? 

If not, this is a good workaround I guess.
